### ISSUE TITLE
Allow LoadControl to prevent loading indefinitely

### ISSUE THREAD

**Issue by wkarl** [google/ExoPlayer#9945]
My use case here is that I would like to prevent loading by overriding DefaultLoadControl if the user switches to a mobile data connection. Due to data caps not everybody wants to use their data towards video streaming. I would like to offer a setting to opt in using mobile data (just like you need to explicitly allow automatic app updates when not on wifi).

You can easily reproduce this by overriding `shouldContinueLoading` and always returning `false`.

Apparently there are some legitimate cases where the player could get stuck which is why there is a check for this in `ExoPlayerImplInternal`:

```
if (!playbackInfo.isLoading
    && playbackInfo.totalBufferedDurationUs < 500_000
    && isLoadingPossible()) {
  // Throw if the LoadControl prevents loading even if the buffer is empty or almost empty. We
  // can't compare against 0 to account for small differences between the renderer position
  // and buffered position in the media at the point where playback gets stuck.
  throw new IllegalStateException("Playback stuck buffering and not loading");
}
```

Is there a better way to achieve this without running into an Exception? I would like this to be fully recoverable: if the user connects to wifi again streaming should continue without having to prepare the player again.

If not I would like to request a configuration flag to override this for my use case. I cannot imagine that I am the only one facing this.

**Comment by tonihei** [comment]
>  I would like this to be fully recoverable: if the user connects to wifi again streaming should continue without having to prepare the player again.

This is an interesting request in that the player is technically doing the right thing: it detects the playback as being stuck and fails. The issue is that you actually want the player to be stuck forever (or until the WiFi connection comes back). I think it would make more sense to let the player fail, display an appropriate message to the user ("no internet connection / waiting for WiFi") and just call `player.prepare()` once the Wifi is available again. There isn't a big cost involved in calling `prepare` again (other than recreating the decoder) because there is no already downloaded data left to play anyway, so it's unclear why you want to avoid re-preparation. 

Two further thoughts on general setup:
 - It's not too unlikely that the user will stay on the mobile data connection for a considerable amount of time, so they might not expect playback to continue anyway? 
 - To create a consistent user experience, you likely need to implement the same logic for a client that goes offline completely. However, that's not easily possible because sudden switches to offline will almost certainly result in network-related playback errors instead of a stuck-buffering exception, so it's unlikely you can prevent the player from going into an error state when the connection is lost completely.

**Comment by wkarl** [comment]
Thank you for your detailed response @tonihei 
About the two further thoughts you mentioned:
> It's not too unlikely that the user will stay on the mobile data connection for a considerable amount of time, so they might not expect playback to continue anyway?

This only applies to users with a very limited data plan. Many users will want to continue playing on mobile data. That's why there is a "allow mobile data playback" button in the dialog we display: it will dismiss the message and continue playing immediately (in addition to a switch in the app settings).

> To create a consistent user experience, you likely need to implement the same logic for a client that goes offline completely. However, that's not easily possible because sudden switches to offline will almost certainly result in network-related playback errors instead of a stuck-buffering exception, so it's unlikely you can prevent the player from going into an error state when the connection is lost completely.

You'd be surprised how well this works. I modified `DefaultLoadErrorHandlingPolicy` to keep retrying playback for this case. As soon as a connection is restored the stream continues playing without fail.

As you suggested I currently just call `player.prepare()` again but there are a few downsides:
- The error must be suppressed for this specific case when reporting errors to Firebase etc. to avoid inflating playback errors.
- Semantics: I see this specific case as more of a prompt than an actual error. Handling in the app is very different to regular errors.
- Recovery would be simpler: just unblock `LoadControl` instead of specialised error handling.
- To make sure the video is resumed at the correct position, I either need to remember the position where the error occurred or skip setting the media item again (or use `player.setMediaItem(mediaItem, false)` to keep the position).

Despite the successful workaround this would definitely be a nice-to-have for me.

**Comment by tonihei** [comment]
> To make sure the video is resumed at the correct position, I either need to remember the position where the error occurred or skip setting the media item again (or use player.setMediaItem(mediaItem, false) to keep the position).

I believe this should just work without any special workarounds? If you just call `prepare()` after the error and nothing else, playback should continue from where it left off and the whole playlist should still be intact.

>  You'd be surprised how well this works. I modified DefaultLoadErrorHandlingPolicy to keep retrying playback for this case. As soon as a connection is restored the stream continues playing without fail.

This is certainly possible (although I think not the original intention of the `LoadErrorHandlingPolicy`). If you handle network errors in this way anyway, I believe you can achieve your goal with the mobile data connection via the same path by using a `DataSource` wrapper that throws a custom exception when Wifi is lost.

Something like (pseudo-code):
```
class NetworkAwareDataSource implements DataSource {

   private DataSource actualDataSource;

   @Override 
   public void allDataSourceMethods(...) {
      actualDataSource.method(...);
    }

   @Override 
   public int read(....) throws IOException {
      if (networkLost) {
        throw new CustomNetworkLostIOException();
      }
      return actualDataSource.read(...);
   }
}
```
and you can then pass this `DataSource` to the `DefaultMediaSourceFactory` in the `ExoPlayer.Builder` (or your more specific `MediaSource.Factory` depending on your setup).

**Comment by wkarl** [comment]
That looks like a good approach, thank you for the suggestion!

**Comment by tonihei** [comment]
Closing the issue, because the question was answered.
